alicloud/resource_alicloud_snapshot_policy.go

```diff
@@ -0,0 +1,151 @@
+package alicloud
+
+import (
+	"github.com/aliyun/alibaba-cloud-sdk-go/sdk/requests"
+	"github.com/aliyun/alibaba-cloud-sdk-go/services/ecs"
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/terraform-providers/terraform-provider-alicloud/alicloud/connectivity"
+	"time"
+)
+
+func resourceAliyunSnapshotPolicy() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAliyunSnapshotPolicyCreate,
+		Read:   resourceAliyunSnapshotPolicyRead,
+		Update: resourceAliyunSnapshotPolicyUpdate,
+		Delete: resourceAliyunSnapshotPolicyDelete,
+
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"repeat_weekdays": {
+				Type:     schema.TypeSet,
+				Required: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+			},
+			"retention_days": {
+				Type:     schema.TypeInt,
+				Required: true,
+			},
+			"time_points": {
+				Type:     schema.TypeSet,
+				Required: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+			},
+		},
+	}
+}
+
+func resourceAliyunSnapshotPolicyCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*connectivity.AliyunClient)
+
+	args := ecs.CreateCreateAutoSnapshotPolicyRequest()
+	args.AutoSnapshotPolicyName = d.Get("name").(string)
+	args.RepeatWeekdays = convertListToJsonString(d.Get("repeat_weekdays").(*schema.Set).List())
+	args.RetentionDays = requests.NewInteger(d.Get("retention_days").(int))
+	args.TimePoints = convertListToJsonString(d.Get("time_points").(*schema.Set).List())
+
+	raw, err := client.WithEcsClient(func(ecsClient *ecs.Client) (interface{}, error) {
+		return ecsClient.CreateAutoSnapshotPolicy(args)
+	})
+	if err != nil {
+		return WrapErrorf(err, DefaultErrorMsg, "alicloud_snapshot_policy", AlibabaCloudSdkGoERROR)
+	}
+	resp := raw.(*ecs.CreateAutoSnapshotPolicyResponse)
+	d.SetId(resp.AutoSnapshotPolicyId)
+
+	ecsService := EcsService{client}
+	if err := ecsService.WaitForSnapshotPolicy(d.Id(), SnapshotPolicyNormal, DefaultTimeout); err != nil {
+		return WrapError(err)
+	}
+
+	return resourceAliyunSnapshotPolicyRead(d, meta)
+}
+
+func resourceAliyunSnapshotPolicyRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*connectivity.AliyunClient)
+	ecsService := EcsService{client}
+	snapshotPolicy, err := ecsService.DescribeSnapshotPolicy(d.Id())
+	if err != nil {
+		if NotFoundError(err) {
+			d.SetId("")
+			return nil
+		}
+		return WrapError(err)
+	}
+
+	d.Set("name", snapshotPolicy.AutoSnapshotPolicyName)
+	weekdays, err := convertJsonStringToList(snapshotPolicy.RepeatWeekdays)
+	if err != nil {
+		return WrapErrorf(err, DefaultErrorMsg, "alicloud_snapshot_policy", ProviderERROR)
```

Comment:
The line 87 error is a service error or common error, just `WrapError(err)` is ok.